# Introduction/Business Problem

New York City and the city of Toronto are very diverse and are the financial capitals of their respective countries. Since we explored certain boroughs of  New York City and the city of Toronto and segmented and clustered their neighborhoods, lets try to cluster whole city and their entire neighborhoods to find similarities among them.

The aim of this project is to build a guide which will show which neighborhoods are similar between these two cities. This can be helpful for businneses that are trying to expand from one city to another or for people trying to relocate from one city to another and doesnt want to miss out the amenities thay had in the previous city. 

#Data

##New york data

In [ ]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
import json # library to handle JSON files
from pandas import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


New york has a total of 5 boroughs and 306 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 5 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood. 


In [ ]:
!wget -q -O 'newyork_data.json' https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs/newyork_data.json
print('Data downloaded!')

Data downloaded!


#### Load and explore the data


In [ ]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [ ]:
newyork_data

{'bbox': [-74.2492599487305,
  40.5033187866211,
  -73.7061614990234,
  40.9105606079102],
 'crs': {'properties': {'name': 'urn:ogc:def:crs:EPSG::4326'}, 'type': 'name'},
 'features': [{'geometry': {'coordinates': [-73.84720052054902,
     40.89470517661],
    'type': 'Point'},
   'geometry_name': 'geom',
   'id': 'nyu_2451_34572.1',
   'properties': {'annoangle': 0.0,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661],
    'borough': 'Bronx',
    'name': 'Wakefield',
    'stacked': 1},
   'type': 'Feature'},
  {'geometry': {'coordinates': [-73.82993910812398, 40.87429419303012],
    'type': 'Point'},
   'geometry_name': 'geom',
   'id': 'nyu_2451_34572.2',
   'properties': {'annoangle': 0.0,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.874294193

In [ ]:
#let's define a new variable that includes this data  in the features key, which is basically a list of the neighborhoods.
neighborhoods_data = newyork_data['features']

In [ ]:
neighborhoods_data[0]

{'geometry': {'coordinates': [-73.84720052054902, 40.89470517661],
  'type': 'Point'},
 'geometry_name': 'geom',
 'id': 'nyu_2451_34572.1',
 'properties': {'annoangle': 0.0,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661],
  'borough': 'Bronx',
  'name': 'Wakefield',
  'stacked': 1},
 'type': 'Feature'}

#### Tranform the data into a _pandas_ dataframe


In [ ]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [ ]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.


In [ ]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Then let's loop through the data and fill the dataframe one row at a time.


In [ ]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [ ]:
#Lets check that the dataset has all 5 boroughs and 306 neighborhoods.
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


#### Use geopy library to get the latitude and longitude values of New York City.


In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.


In [ ]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


#### Create a map of New York with neighborhoods superimposed on top.


In [ ]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

##Toronto data

###Web Scraping

Toronto has a total of 10 boroughs and 103 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 10 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood. 

We will be using the following source to acquire the data about neighbourhoods in Toronto

Source: [https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M)

In [ ]:
url ="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [ ]:
df = pd.read_html(url,header=0)
df

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 5           M6A        North York   
 6           M7A  Downtown Toronto   
 7           M8A      Not assigned   
 8           M9A         Etobicoke   
 9           M1B       Scarborough   
 10          M2B      Not assigned   
 11          M3B        North York   
 12          M4B         East York   
 13          M5B  Downtown Toronto   
 14          M6B        North York   
 15          M7B      Not assigned   
 16          M8B      Not assigned   
 17          M9B         Etobicoke   
 18          M1C       Scarborough   
 19          M2C      Not assigned   
 20          M3C        North York   
 21          M4C         East York   
 22          M5C  Downtown Toronto   
 23          M6C              York   
 24          M7C      Not assigned   
 25         

In [ ]:
#no of tables in  the source page
len(df)

3

In [ ]:
#so there are 3 tables, lets select the 1st table
df = df[0]
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

In [ ]:
#dropping cells with a borough that is Not assigned.
df.drop(df.index[df['Borough'] == "Not assigned"], inplace = True)
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Vic

In [ ]:
#assign nan to not assigned, if it exists in Neighbourhood
import numpy as np
df.replace("Not assigned",np.nan,inplace=True)
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Vic

In [ ]:
#checking for null values
df.isnull().sum()

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

Hence there are no null values

In [ ]:
df.reset_index(drop=True,inplace=True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Vic

In [ ]:
df.shape

(103, 3)

Source of geographical coordinates of the neighborhoods in toronto:
http://cocl.us/Geospatial_data

In [ ]:
!wget -q http://cocl.us/Geospatial_data

In [ ]:
df1 = pd.read_csv("/content/Geospatial_data")
df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [ ]:
# merge df with df1 to add latitude/longitude for each neighborhood
df_merge = df.join(df1.set_index("Postal Code"),on="Postal Code")
df_merge.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [ ]:
#lets rename our dataframe
neighborhoods1 = df_merge

In [ ]:
#lets drop postal code column and rename neighbourhood to neighborhood
df_merge.drop("Postal Code",axis=1,inplace=True)

In [ ]:
df_merge.rename(columns={"Neighbourhood":"Neighborhood"},inplace=True)

In [ ]:
df_merge.head()

Borough                                 Neighborhood   Latitude  \
0        North York                                    Parkwoods  43.753259   
1        North York                             Victoria Village  43.725882   
2  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
3        North York             Lawrence Manor, Lawrence Heights  43.718518   
4  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [ ]:
#Examining number of boroughs and neighborhoods
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods1['Borough'].unique()),
        neighborhoods1.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto


In [ ]:
address = 'City of Toronto, Ontario ,Canada'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of City of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of City of Toronto are 43.6534817, -79.3839347.


#### Create a map of City of Toronto with neighborhoods superimposed on top.


In [ ]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods1['Latitude'], neighborhoods1['Longitude'], neighborhoods1['Borough'], neighborhoods1['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [ ]:
#boroughs in toronto
neighborhoods1['Borough'].unique()

array(['North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

##Combining both data and creating a new data frame

In [ ]:
#New york neighborhoods
neighborhoods.head(2)

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939

In [ ]:
neighborhoods.shape

(306, 4)

In [ ]:
#Toronto neighborhoods
df_merge.head(2)

Borough      Neighborhood   Latitude  Longitude
0  North York         Parkwoods  43.753259 -79.329656
1  North York  Victoria Village  43.725882 -79.315572

In [ ]:
df_merge.shape

(103, 4)

In [ ]:
#Combined neighborhoods data
neighborhoods =neighborhoods.append(df_merge)

In [ ]:
neighborhoods.shape

(409, 4)

In [ ]:
neighborhoods.reset_index(drop=True,inplace=True)
neighborhoods

Borough                                       Neighborhood  \
0               Bronx                                          Wakefield   
1               Bronx                                         Co-op City   
2               Bronx                                        Eastchester   
3               Bronx                                          Fieldston   
4               Bronx                                          Riverdale   
5               Bronx                                        Kingsbridge   
6           Manhattan                                        Marble Hill   
7               Bronx                                           Woodlawn   
8               Bronx                                            Norwood   
9               Bronx                                     Williamsbridge   
10              Bronx                                         Baychester   
11              Bronx                                     Pelham Parkway   
12              Bronx                                        City Island   
13              Bronx                                       Bedford Park   
14              Bronx                                 University Heights   
15              Bronx                                     Morris Heights   
16              Bronx                                            Fordham   
17              Bronx                                       East Tremont   
18              Bronx                                         West Farms   
19              Bronx                                       High  Bridge   
20              Bronx                                            Melrose   
21              Bronx                                         Mott Haven   
22              Bronx                                        Port Morris   
23              Bronx                                           Longwood   
24              Bronx                                        Hunts Point   
25              Bronx                                         Morrisania   
26              Bronx                                          Soundview   
27              Bronx                                       Clason Point   
28              Bronx                                        Throgs Neck   
29              Bronx                                       Country Club   
30              Bronx                                        Parkchester   
31              Bronx                                 Westchester Square   
32              Bronx                                           Van Nest   
33              Bronx                                        Morris Park   
34              Bronx                                            Belmont   
35              Bronx                                     Spuyten Duyvil   
36              Bronx                                    North Riverdale   
37              Bronx                                         Pelham Bay   
38              Bronx                                      Schuylerville   
39              Bronx                                     Edgewater Park   
40              Bronx                                        Castle Hill   
41              Bronx                                          Olinville   
42              Bronx                                     Pelham Gardens   
43              Bronx                                          Concourse   
44              Bronx                                          Unionport   
45              Bronx                                           Edenwald   
46           Brooklyn                                          Bay Ridge   
47           Brooklyn                                        Bensonhurst   
48           Brooklyn                                        Sunset Park   
49           Brooklyn                                         Greenpoint   
50           Brooklyn                                          Gravesend   
51           Brooklyn                                     Brig

In [ ]:
neighborhoods['Borough'].unique()

array(['Bronx', 'Manhattan', 'Brooklyn', 'Queens', 'Staten Island',
       'North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

In [ ]:
# create visualize both data
map_combined_cities = folium.Map(location=[latitude, longitude], zoom_start=6)

# add markers to map
for lat, lng, label in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_combined_cities)  
    
map_combined_cities

We can zoom into whichever city we are intersted in and see the neighborhoods

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.


#### Define Foursquare Credentials and Version


In [ ]:
CLIENT_ID = 'VN313PRPAGZCC510RQQTBYUPMGMLUUHEXXZ0LGWYBU2440H1' # your Foursquare ID
CLIENT_SECRET = 'YB12DGTPUO2UM2YP4CAYC1HNB4NOIWLAYM1NYE3HIPBUADYC' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VN313PRPAGZCC510RQQTBYUPMGMLUUHEXXZ0LGWYBU2440H1
CLIENT_SECRET:YB12DGTPUO2UM2YP4CAYC1HNB4NOIWLAYM1NYE3HIPBUADYC


#### Let's create a function to get all venues in the neighborhoods 


In [ ]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [ ]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [ ]:
venues_list = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Marble Hill
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker

In [ ]:
print(venues_list.shape)
venues_list.head()

(12275, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

              Venue  Venue Latitude  Venue Longitude  Venue Category  
0  Lollipops Gelato       40.894123       -73.845892    Dessert Shop  
1          Rite Aid       40.896649       -73.844846        Pharmacy  
2         Walgreens       40.896528       -73.844700        Pharmacy  
3  Carvel Ice Cream       40.890487       -73.848568  Ice Cream Shop  
4           Dunkin'       40.890459       -73.849089      Donut Shop

Let's check how many venues were returned for each neighborhood


In [ ]:
venues_list.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  9   
Allerton                                                               30   
Annadale                                                                9   
Arden Heights                                                           4   
Arlington                                                               5   
Arrochar                                                               22   
Arverne                                                                20   
Astoria                                                                96   
Astoria Heights                                                        15   
Auburndale                                                             19   
Bath Beach                                                             45   
Bathurst Manor, Wilson Heights, Downsview North                        21   
Battery Park City                                                      79   
Bay Ridge                                                              80   
Bay Terrace                                                            45   
Baychester                                                             22   
Bayside                                                                70   
Bayswater                                                               2   
Bayview Village                                                         4   
Bedford Park                                                           38   
Bedford Park, Lawrence Manor East                                      24   
Bedford Stuyvesant                                                     28   
Beechhurst                                                             15   
Bellaire                                                               12   
Belle Harbor                                                           16   
Bellerose                                                              21   
Belmont                                                               100   
Bensonhurst                                                            34   
Berczy Park                                                            61   
Bergen Beach                                                            6   
Birch Cliff, Cliffside West                                             4   
Blissville                                                             21   
Bloomfield                                                              4   
Boerum Hill                                                            89   
Borough Park                                                           19   
Breezy Point                                                            6   
Briarwood                                                              11   
Brighton Beach                                                         42   
Broad Channel                                                           5   
Broadway Junction                                                      16   
Brockton, Parkdale Village, Exhibition Place                           22   
Bronxdale                                                              13   
Brooklyn Heights                                                      100   
Brookville                                                              1   
Brownsville                                                            14   
Bulls Head                                                             44   
Bushwick                                                               73   
Business reply mail Processing Centre, South Ce...                     16   
Butler Manor                                                            5   
CN Tower, King and Spadina, Railway Lands, Harb.

In [ ]:
venues_list.shape

(12275, 7)

#### Let's find out how many unique categories can be curated from all the returned venues


In [ ]:
print('There are {} uniques categories.'.format(len(venues_list['Venue Category'].unique())))

There are 464 uniques categories.


<a id='item3'></a>


## Analyze Each Neighborhood


In [ ]:
# one hot encoding
combined_cities_onehot = pd.get_dummies(venues_list[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
combined_cities_onehot['Neighborhood'] = venues_list['Neighborhood'] 

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  \
0            0                  0               0                  0   
1            0                  0               0                  0   
2            0                  0               0                  0   
3            0                  0               0                  0   
4            0                  0               0                  0   

   African Restaurant  Airport  Airport Food Court  Airport Gate  \
0                   0        0                   0             0   
1                   0        0                   0             0   
2                   0        0                   0             0   
3                   0        0                   0             0   
4                   0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Animal Shelter  Antique Shop  Aquarium  Arcade  Arepa Restaurant  \
0               0             0         0       0                 0   
1               0             0         0       0                 0   
2               0             0         0       0                 0   
3               0             0         0       0                 0   
4               0             0         0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  Auditorium  \
0                     0                 0                   0           0   
1                     0                 0                   0           0   
2                     0                 0                   0           0   
3                     0                 0                   0           0   
4                     0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Garage  Auto Workshop  \
0                      0                    0            0              0   
1                      0                    0            0              0   
2                      0                    0            0              0   
3                      0                    0            0              0   
4                      0                    0            0              0   

   Automotive Shop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                0          0           0           0       0     0    0   
1                0          0           0           0       0     0    0   
2                0          0           0           0       0     0    0   
3                0          0           0           0       0     0    0   
4                0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 

In [ ]:
combined_cities_onehot.shape

(12275, 464)

In [ ]:
combined_cities_onehot.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12275 entries, 0 to 12274
Columns: 464 entries, Women's Store to Wings Joint
dtypes: object(1), uint8(463)
memory usage: 5.5+ MB


#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [ ]:
combined_cities_grouped = combined_cities_onehot.groupby('Neighborhood').mean().reset_index()
combined_cities_grouped

In [ ]:
combined_cities_grouped.shape

(394, 464)

#### Let's print each neighborhood along with the top 5 most common venues


In [ ]:
num_top_venues = 5

for hood in combined_cities_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = combined_cities_grouped[combined_cities_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.25
1             Breakfast Spot  0.25
2                 Print Shop  0.25
3                     Lounge  0.25
4              Women's Store  0.00


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place  0.22
1         Coffee Shop  0.11
2        Skating Rink  0.11
3                 Gym  0.11
4  Athletics & Sports  0.11


----Allerton----
                venue  freq
0         Pizza Place  0.13
1       Deli / Bodega  0.10
2         Supermarket  0.10
3  Chinese Restaurant  0.07
4                 Spa  0.07


----Annadale----
            venue  freq
0     Pizza Place  0.22
1  Cosmetics Shop  0.11
2            Food  0.11
3    Liquor Store  0.11
4   Train Station  0.11


----Arden Heights----
           venue  freq
0       Pharmacy  0.25
1  Deli / Bodega  0.25
2    Coffee Shop  0.25
3    Pizza Place  0.25
4  Women's Store  0.00


----Arlington----
           venue  freq
0       Bus Stop

#### Let's put that into a _pandas_ dataframe


First, let's write a function to sort the venues in descending order.


In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = combined_cities_grouped['Neighborhood']

for ind in np.arange(combined_cities_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(combined_cities_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0               Agincourt            Print Shop        Breakfast Spot   
1  Alderwood, Long Branch           Pizza Place    Athletics & Sports   
2                Allerton           Pizza Place           Supermarket   
3                Annadale           Pizza Place                   Pub   
4           Arden Heights           Coffee Shop              Pharmacy   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Latin American Restaurant                Lounge           Wings Joint   
1               Dance Studio                   Pub                   Gym   
2              Deli / Bodega                   Spa    Chinese Restaurant   
3                      Diner            Restaurant         Train Station   
4              Deli / Bodega           Pizza Place           Wings Joint   

         6th Most Common Venue        7th Most Common Venue  \
0           Falafel Restaurant  Eastern European Restaurant   
1                 Skating Rink               Sandwich Place   
2                 Intersection            Electronics Store   
3                         Food               Cosmetics Shop   
4  Eastern European Restaurant            Electronics Store   

  8th Most Common Venue  9th Most Common Venue       10th Most Common Venue  
0     Electronics Store    Empanada Restaurant        Entertainment Service  
1           Coffee Shop          Event Service         Ethiopian Restaurant  
2         Grocery Store         Breakfast Spot        Check Cashing Service  
3          Liquor Store         Duty-free Shop  Eastern European Restaurant  
4   Empanada Restaurant  Entertainment Service                  Escape Room

In [ ]:
neighborhoods_venues_sorted.shape

(394, 11)

<a id='item4'></a>


## Cluster Neighborhoods


Lets run _k_-means to cluster the neighborhood into 10 clusters, so that we can have more choice in identifying clusters that are very similar


In [ ]:
# set number of clusters
kclusters = 10

combined_cities_grouped_clustering = combined_cities_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(combined_cities_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 1, 1, 1, 1, 1, 0, 0, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [ ]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
neighborhoods_venues_sorted = neighborhoods_venues_sorted.join(neighborhoods.set_index('Neighborhood'), on='Neighborhood')

neighborhoods_venues_sorted.head() # check the last columns!

Cluster Labels            Neighborhood 1st Most Common Venue  \
0               0               Agincourt            Print Shop   
1               1  Alderwood, Long Branch           Pizza Place   
2               1                Allerton           Pizza Place   
3               1                Annadale           Pizza Place   
4               1           Arden Heights           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant                Lounge   
1    Athletics & Sports               Dance Studio                   Pub   
2           Supermarket              Deli / Bodega                   Spa   
3                   Pub                      Diner            Restaurant   
4              Pharmacy              Deli / Bodega           Pizza Place   

  5th Most Common Venue        6th Most Common Venue  \
0           Wings Joint           Falafel Restaurant   
1                   Gym                 Skating Rink   
2    Chinese Restaurant                 Intersection   
3         Train Station                         Food   
4           Wings Joint  Eastern European Restaurant   

         7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0  Eastern European Restaurant     Electronics Store    Empanada Restaurant   
1               Sandwich Place           Coffee Shop          Event Service   
2            Electronics Store         Grocery Store         Breakfast Spot   
3               Cosmetics Shop          Liquor Store         Duty-free Shop   
4            Electronics Store   Empanada Restaurant  Entertainment Service   

        10th Most Common Venue        Borough   Latitude  Longitude  
0        Entertainment Service    Scarborough  43.794200 -79.262029  
1         Ethiopian Restaurant      Etobicoke  43.602414 -79.543484  
2        Check Cashing Service          Bronx  40.865788 -73.859319  
3  Eastern European Restaurant  Staten Island  40.538114 -74.178549  
4                  Escape Room  Staten Island  40.549286 -74.185887

In [ ]:
neighborhoods_venues_sorted.shape

(403, 15)

Finally, let's visualize the resulting clusters


In [ ]:
neighborhoods_venues_sorted['Cluster Labels'].unique()

array([0, 1, 8, 5, 6, 3, 4, 2, 7, 9], dtype=int32)

In [ ]:
neighborhoods_venues_sorted['Cluster Labels'].value_counts()

0    167
1    160
5     57
6     10
9      2
8      2
4      2
7      1
3      1
2      1
Name: Cluster Labels, dtype: int64

>As we can seethere are two rows that are labeled nan, these can be outliers as they cannot be assigned to any group

In [ ]:
#removing rows with nan values
north_york_merged.dropna(axis=0,inplace=True)

In [ ]:
print("labels =", north_york_merged['Cluster Labels'].unique())
north_york_merged['Cluster Labels'].value_counts()

labels = [3. 0. 4. 1. 2.]


0.0    18
2.0     1
1.0     1
4.0     1
3.0     1
Name: Cluster Labels, dtype: int64

In [ ]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=6)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(neighborhoods_venues_sorted['Latitude'], neighborhoods_venues_sorted['Longitude'], neighborhoods_venues_sorted['Neighborhood'], neighborhoods_venues_sorted['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>


##Examine Clusters


Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.


#### Cluster 1


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 0, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood  \
0                                            Agincourt   
7                                              Arverne   
8                                              Astoria   
10                                          Auburndale   
13                                   Battery Park City   
14                                           Bay Ridge   
15                                         Bay Terrace   
15                                         Bay Terrace   
17                                             Bayside   
19                                     Bayview Village   
21                   Bedford Park, Lawrence Manor East   
22                                  Bedford Stuyvesant   
28                                         Bensonhurst   
29                                         Berczy Park   
31                         Birch Cliff, Cliffside West   
34                                         Boerum Hill   
38                                      Brighton Beach   
40                                   Broadway Junction   
41        Brockton, Parkdale Village, Exhibition Place   
43                                    Brooklyn Heights   
47                                            Bushwick   
48   Business reply mail Processing Centre, South C...   
50   CN Tower, King and Spadina, Railway Lands, Har...   
52                                     Cambria Heights   
53               Canada Post Gateway Processing Centre   
55                                       Carnegie Hill   
56                                     Carroll Gardens   
59                                           Cedarbrae   
60                                  Central Bay Street   
61                                      Central Harlem   
62                                          Charleston   
63                                             Chelsea   
63                                             Chelsea   
64                                           Chinatown   
65                                            Christie   
66                                Church and Wellesley   
67                                         City Island   
69                                        Civic Center   
73     Cliffside, Cliffcrest, Scarborough Village West   
75                                             Clinton   
76                                        Clinton Hill   
78                                         Cobble Hill   
80                      Commerce Court, Victoria Hotel   
81                                             Concord   
89                                          Davisville   
91    Del Ray, Mount Dennis, Keelsdale and Silverthorn   
93                                           Don Mills   
93                                           Don Mills   
95   Dorset Park, Wexford Heights, Scarborough Town...   
98                                            Downtown   
99                        Dufferin, Dovercourt Village   
100                                              Dumbo   
104                                        East Harlem   
108                                       East Village   
109                                  East Williamsburg   
113                                     Edgewater Park   
114                                        Egbertville   
116                                           Elmhurst   
120  Eringate, Bloordale Gardens, Old Burnhamthorpe...   
121                       Fairview, Henry Farm, Oriole   
124                                 Financial District   
125             First Canadian Place, Underground city   
127                                           Flatiron   
130                                           Flushing   
135                                        Fort Greene   
139                                       Fulton Ferry   
140                           Garden District, Ryerson   
141                                         Georgetown   
142                                    G

#### Cluster 2


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 1, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood  \
1                               Alderwood, Long Branch   
2                                             Allerton   
3                                             Annadale   
4                                        Arden Heights   
5                                            Arlington   
6                                             Arrochar   
9                                      Astoria Heights   
11                                          Bath Beach   
12     Bathurst Manor, Wilson Heights, Downsview North   
16                                          Baychester   
20                                        Bedford Park   
23                                          Beechhurst   
24                                            Bellaire   
26                                           Bellerose   
27                                             Belmont   
32                                          Blissville   
35                                        Borough Park   
37                                           Briarwood   
39                                       Broad Channel   
42                                           Bronxdale   
44                                          Brookville   
46                                          Bulls Head   
54                                            Canarsie   
57                                         Castle Hill   
58                                   Castleton Corners   
68                                           City Line   
70                                   Claremont Village   
71             Clarks Corners, Tam O'Shanter, Sullivan   
74                                             Clifton   
77                                          Co-op City   
79                                       College Point   
82                                           Concourse   
83                                   Concourse Village   
84                                        Coney Island   
85                                              Corona   
87                                       Crown Heights   
88                                       Cypress Hills   
92                                         Ditmas Park   
94                                        Dongan Hills   
96                                          Douglaston   
102                                      East Elmhurst   
103                                      East Flatbush   
105                                      East New York   
107                                       East Tremont   
110                                        Eastchester   
112                                           Edgemere   
115                                           Elm Park   
117                                        Eltingville   
119                                            Erasmus   
122                                       Far Rockaway   
126                                           Flatbush   
128                                          Flatlands   
129                                        Floral Park   
131                                            Fordham   
133                                       Forest Hills   
134                               Forest Hills Gardens   
136                                      Fort Hamilton   
138                                      Fresh Meadows   
143                                          Glen Oaks   
144                                          Glencairn   
145                                           Glendale   
151                                           Grasmere   
152                                          Gravesend   
153                                        Great Kills   
155                                         Greenridge   
157                                        Grymes Hill   
162                                  Heartland Village   
163                                       High  Bridge   
165                                     

#### Cluster 3


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 2, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
303     Roselawn                Garden           Wings Joint   

    3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
303        Farmers Market  Eastern European Restaurant     Electronics Store   

    6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
303   Empanada Restaurant  Entertainment Service           Escape Room   

    9th Most Common Venue 10th Most Common Venue  
303  Ethiopian Restaurant          Event Service

#### Cluster 4


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 3, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
175  Humberlea, Emery        Baseball Field           Wings Joint   

    3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
175                  Farm  Eastern European Restaurant     Electronics Store   

    6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
175   Empanada Restaurant  Entertainment Service           Escape Room   

    9th Most Common Venue 10th Most Common Venue  
175  Ethiopian Restaurant          Event Service

#### Cluster 5


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 4, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood 1st Most Common Venue  \
183                                    Jamaica Estates          Intersection   
228  Milliken, Agincourt North, Steeles East, L'Amo...          Intersection   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
183               Dog Run                  Farm  Eastern European Restaurant   
228                  Park            Playground                         Farm   

           5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
183            Electronics Store   Empanada Restaurant  Entertainment Service   
228  Eastern European Restaurant     Electronics Store    Empanada Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
183            Escape Room  Ethiopian Restaurant          Event Service  
228  Entertainment Service           Escape Room   Ethiopian Restaurant

#### Cluster 6


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 5, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood  \
25                                        Belle Harbor   
30                                        Bergen Beach   
33                                          Bloomfield   
36                                        Breezy Point   
45                                         Brownsville   
49                                        Butler Manor   
86                                        Country Club   
90                                    Davisville North   
97                                           Downsview   
97                                           Downsview   
97                                           Downsview   
97                                           Downsview   
101                                      Dyker Heights   
111                                           Edenwald   
118                                       Emerson Hill   
123                                          Fieldston   
132    Forest Hill North & West, Forest Hill Road Park   
137                                          Fox Hills   
146                    Golden Mile, Clairlea, Oakridge   
149                                       Graniteville   
160                                            Hammels   
167                                  Hillcrest Village   
168                                             Hollis   
169                                         Holliswood   
176                                 Humewood-Cedarvale   
192  Kingsview Village, St. Phillips, Martin Grove ...   
193                                          Laurelton   
209                                              Malba   
211                                    Manhattan Beach   
222                                      Midland Beach   
227                                        Mill Island   
230                        Moore Park, Summerhill East   
239                                           Neponsit   
240                                       New Brighton   
248           North Park, Maple Leaf Park, Upwood Park   
259  Old Mill South, King's Mill Park, Sunnylea, Hu...   
263                                    Paerdegat Basin   
264                                          Park Hill   
282                                     Queens Village   
284                                       Queensbridge   
285                                      Randall Manor   
296                                          Riverdale   
298                                     Rockaway Beach   
302                                           Rosedale   
302                                           Rosedale   
306                                            Roxbury   
309                      Runnymede, The Junction North   
310                                       Sandy Ground   
313                                           Sea Gate   
320                                        South Beach   
322                                   South Ozone Park   
325                                Springfield Gardens   
344                                        The Beaches   
352                                        Tottenville   
353                                             Travis   
371                                         West Farms   
387                                   Woodbine Heights   

          1st Most Common Venue      2nd Most Common Venue  \
25                        Beach                        Pub   
30              Harbor / Marina                       Park   
33                   Theme Park                   Bus Stop   
36                        Beach        Monument / Landmark   
45                Moving Target        Fried Chicken Joint   
49                         Pool             Baseball Field   
86               Sandwich Place         Athletics & Sports   
90            Food & Drink Shop               Dance Studio   
97                Grocery Store                       Park   
97                Grocery Store                       Par

#### Cluster 7


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 6, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood 1st Most Common Venue  \
51                             Caledonia-Fairbanks                  Park   
72                                    Clason Point                  Park   
106  East Toronto, Broadview North (Old East York)          Intersection   
195                                  Lawrence Park                  Park   
269                                      Parkwoods                  Park   
318                                     Somerville                  Park   
346  The Kingsway, Montgomery Road, Old Mill North                  Park   
349                                      Todt Hill                  Park   
376                                         Weston     Convenience Store   
392                                York Mills West     Convenience Store   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
51          Women's Store                  Pool                         Farm   
72                   Pool         Boat or Ferry    South American Restaurant   
106                  Park     Convenience Store            Accessories Store   
195              Bus Line           Swim School  Eastern European Restaurant   
269     Food & Drink Shop           Wings Joint                         Farm   
318           Wings Joint                  Farm  Eastern European Restaurant   
346                 River           Wings Joint                         Farm   
349           Wings Joint                  Farm  Eastern European Restaurant   
376                  Park                  Farm  Eastern European Restaurant   
392                  Park                  Farm  Eastern European Restaurant   

           5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
51   Eastern European Restaurant     Electronics Store    Empanada Restaurant   
72                      Bus Stop         Grocery Store      Convenience Store   
106  Eastern European Restaurant     Electronics Store    Empanada Restaurant   
195            Electronics Store   Empanada Restaurant  Entertainment Service   
269  Eastern European Restaurant     Electronics Store    Empanada Restaurant   
318            Electronics Store   Empanada Restaurant  Entertainment Service   
346  Eastern European Restaurant     Electronics Store    Empanada Restaurant   
349            Electronics Store   Empanada Restaurant  Entertainment Service   
376            Electronics Store   Empanada Restaurant  Entertainment Service   
392            Electronics Store   Empanada Restaurant  Entertainment Service   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
51   Entertainment Service           Escape Room   Ethiopian Restaurant  
72             Curling Ice   Empanada Restaurant           Cupcake Shop  
106  Entertainment Service           Escape Room   Ethiopian Restaurant  
195            Escape Room  Ethiopian Restaurant          Event Service  
269  Entertainment Service           Escape Room   Ethiopian Restaurant  
318            Escape Room  Ethiopian Restaurant          Event Service  
346  Entertainment Service           Escape Room   Ethiopian Restaurant  
349            Escape Room  Ethiopian Restaurant          Event Service  
376            Escape Room  Ethiopian Restaurant          Event Service  
392            Escape Room  Ethiopian Restaurant          Event Service

#### Cluster 8


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 7, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood 1st Most Common Venue  \
305  Rouge Hill, Port Union, Highland Creek                   Bar   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
305           Wings Joint                  Farm  Eastern European Restaurant   

    5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
305     Electronics Store   Empanada Restaurant  Entertainment Service   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
305           Escape Room  Ethiopian Restaurant          Event Service

#### Cluster 9


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 8, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
18             Bayswater          Tennis Court            Playground   
311  Scarborough Village            Playground           Wings Joint   

    3rd Most Common Venue        4th Most Common Venue  \
18            Wings Joint                         Farm   
311        Farmers Market  Eastern European Restaurant   

           5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
18   Eastern European Restaurant     Electronics Store    Empanada Restaurant   
311            Electronics Store   Empanada Restaurant  Entertainment Service   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18   Entertainment Service           Escape Room   Ethiopian Restaurant  
311            Escape Room  Ethiopian Restaurant          Event Service

#### Cluster 10


In [ ]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Cluster Labels'] == 9, neighborhoods_venues_sorted.columns[[1] + list(range(2,12))]]

Neighborhood  \
370  West Deane Park, Princess Gardens, Martin Grov...   
377                                  Wexford, Maryvale   

         1st Most Common Venue 2nd Most Common Venue  \
370  Middle Eastern Restaurant        Farmers Market   
377                Auto Garage                Bakery   

           3rd Most Common Venue      4th Most Common Venue  \
370  Eastern European Restaurant          Electronics Store   
377               Sandwich Place  Middle Eastern Restaurant   

    5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
370   Empanada Restaurant  Entertainment Service           Escape Room   
377           Wings Joint                   Farm     Electronics Store   

    8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
370  Ethiopian Restaurant          Event Service            Event Space  
377   Empanada Restaurant  Entertainment Service            Escape Room

In [ ]:
#Number of neihbor hoods in each clusters
neighborhoods_venues_sorted['Cluster Labels'].value_counts()

0    167
1    160
5     57
6     10
9      2
8      2
4      2
7      1
3      1
2      1
Name: Cluster Labels, dtype: int64

## Results and discussion

- Clusters 3,4,8 are unique to their city as they have only one neighborhood
- Clusters 5,9 has two neighborhoods,each one belonging to different cities.
- Cluster 10 has two neighborhoods both belonging to Toronto
- Cluster 7 has 10 neighborhoods, of which some belong to New York city and the rest belonging to Toronto city  
- Cluster 6 has 57 neighborhoods, of which some belong to New York city and the rest belonging to Toronto city 
- Cluster 2 has 160 neighborhoods, of which some belong to New York city and the rest belonging to Toronto city 
- Cluster 1 has 167 neighborhoods(the maximum), of which some belong to New York city and the rest belonging to Toronto city 

As we can see most of the neighbourhoods fall into cluster 1 and 2 due to the high frequency of venues like coffee shops,cafe,gym, parks, bakery, lounges etc.

These clusters can be used by the people relocating from one city to another city, to find the neighborhoobs with similar amenities to their previous city.

They can be also used for creating business oppurtunities like finding clusters with low frequency of a certain venue and high polulation density. 


